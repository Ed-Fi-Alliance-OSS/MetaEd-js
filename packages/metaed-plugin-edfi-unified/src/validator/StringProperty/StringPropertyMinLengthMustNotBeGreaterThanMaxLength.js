// @flow
import type { MetaEdEnvironment, ValidationFailure, StringProperty } from '../../../../../packages/metaed-core/index';
import type { StringPropertySourceMap } from '../../../../../packages/metaed-core/src/model/property/StringProperty';
import { asStringProperty } from '../../../../../packages/metaed-core/src/model/property/StringProperty';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.string.forEach(string => {
    const stringProperty: StringProperty = asStringProperty(string);
    const minLength: number = Number.parseInt(stringProperty.minLength || '0', 10);
    const maxLength: number = Number.parseInt(stringProperty.maxLength || '0', 10);
    if (minLength <= maxLength) return;

    failures.push({
      validatorName: 'StringPropertyMinLengthMustNotBeGreaterThanMaxLength',
      category: 'error',
      message: `String Property ${string.metaEdName} has min length greater than max length.`,
      sourceMap: ((string.sourceMap: any): StringPropertySourceMap).minLength,
      fileMap: null,
    });
  });

  return failures;
}
