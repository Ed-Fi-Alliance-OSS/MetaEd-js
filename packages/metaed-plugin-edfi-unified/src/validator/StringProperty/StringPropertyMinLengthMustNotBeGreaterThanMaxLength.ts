import { MetaEdEnvironment, ValidationFailure, StringProperty, StringPropertySourceMap } from 'metaed-core';
import { asStringProperty } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.propertyIndex.string.forEach((string) => {
    const stringProperty: StringProperty = asStringProperty(string);
    const minLength: number = Number.parseInt(stringProperty.minLength || '0', 10);
    const maxLength: number = Number.parseInt(stringProperty.maxLength || '0', 10);
    if (minLength <= maxLength) return;

    failures.push({
      validatorName: 'StringPropertyMinLengthMustNotBeGreaterThanMaxLength',
      category: 'error',
      message: `String Property ${string.metaEdName} has min length greater than max length.`,
      sourceMap: (string.sourceMap as StringPropertySourceMap).minLength,
      fileMap: null,
    });
  });

  return failures;
}
