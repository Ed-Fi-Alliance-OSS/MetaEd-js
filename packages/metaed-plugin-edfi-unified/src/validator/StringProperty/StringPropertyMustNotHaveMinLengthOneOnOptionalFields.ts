import {
  MetaEdEnvironment,
  ValidationFailure,
  StringProperty,
  StringPropertySourceMap,
  versionSatisfies,
  PluginEnvironment,
} from '@edfi/metaed-core';
import { asStringProperty } from '@edfi/metaed-core';

export function validate(metaEd: MetaEdEnvironment): ValidationFailure[] {
  const failures: ValidationFailure[] = [];
  metaEd.propertyIndex.string.forEach((string) => {
    const stringProperty: StringProperty = asStringProperty(string);
    const minLength: number = Number.parseInt(stringProperty.minLength || '0', 10);
    if (minLength === 1 && stringProperty.isOptional) {
      const { targetTechnologyVersion } = metaEd.plugin.get('edfiOdsRelational') as PluginEnvironment;
      failures.push({
        validatorName: 'StringPropertyMustNotHaveMinLengthOneOnOptionalFields',
        category: versionSatisfies(targetTechnologyVersion, '>=7.0.0') ? 'error' : 'warning',
        message: versionSatisfies(targetTechnologyVersion, '>=7.0.0')
          ? `String Property ${string.metaEdName} has min length 1 on optional field.`
          : `String Property ${string.metaEdName} has min length 1 on optional field. This warning will be treated as an error in API version 7.x`,
        sourceMap: (string.sourceMap as StringPropertySourceMap).minLength,
        fileMap: null,
      });
    }
  });

  return failures;
}
