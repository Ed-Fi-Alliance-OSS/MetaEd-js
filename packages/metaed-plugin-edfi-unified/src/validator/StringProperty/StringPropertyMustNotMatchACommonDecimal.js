// @flow
import type { MetaEdEnvironment, ValidationFailure } from '../../../../metaed-core/index';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.string.forEach(string => {
    if (metaEd.entity.sharedDecimal.has(string.metaEdName)) {
      failures.push({
        validatorName: 'StringPropertyMustNotMatchACommonDecimal',
        category: 'error',
        message: `String Property ${string.metaEdName} has the same name as a Common Decimal.`,
        sourceMap: string.sourceMap.metaEdName,
        fileMap: null,
      });
    }
  });

  return failures;
}
