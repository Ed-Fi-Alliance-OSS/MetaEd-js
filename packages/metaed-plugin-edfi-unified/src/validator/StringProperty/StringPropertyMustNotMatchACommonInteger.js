// @flow
import type { MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.string.forEach(string => {
    if (metaEd.entity.sharedInteger.has(string.metaEdName)) {
      failures.push({
        validatorName: 'StringPropertyMustNotMatchACommonInteger',
        category: 'error',
        message: `String Property ${string.metaEdName} has the same name as a Common Integer.`,
        sourceMap: string.sourceMap.metaEdName,
        fileMap: null,
      });
    }
  });

  return failures;
}
