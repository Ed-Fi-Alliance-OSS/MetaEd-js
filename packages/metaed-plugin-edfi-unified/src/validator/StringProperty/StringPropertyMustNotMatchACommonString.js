// @flow
import type { MetaEdEnvironment, ValidationFailure } from '../../../../metaed-core/index';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.string.forEach(string => {
    if (metaEd.entity.sharedString.has(string.metaEdName)) {
      failures.push({
        validatorName: 'StringPropertyMustNotMatchACommonString',
        category: 'error',
        message: `String Property ${string.metaEdName} has the same name as a Common String.`,
        sourceMap: string.sourceMap.metaEdName,
        fileMap: null,
      });
    }
  });

  return failures;
}
