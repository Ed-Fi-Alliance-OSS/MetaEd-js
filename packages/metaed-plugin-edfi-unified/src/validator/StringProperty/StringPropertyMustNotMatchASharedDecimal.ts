import { MetaEdEnvironment, ValidationFailure, ModelBase } from 'metaed-core';
import { getEntityForNamespaces } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.string.forEach(property => {
    const referencedEntity: ModelBase | null = getEntityForNamespaces(
      property.metaEdName,
      [property.namespace, ...property.namespace.dependencies],
      'sharedDecimal',
    );

    if (referencedEntity != null) {
      failures.push({
        validatorName: 'StringPropertyMustNotMatchASharedDecimal',
        category: 'error',
        message: `String Property ${property.metaEdName} has the same name as a Shared Decimal.`,
        sourceMap: property.sourceMap.metaEdName,
        fileMap: null,
      });
    }
  });

  return failures;
}
