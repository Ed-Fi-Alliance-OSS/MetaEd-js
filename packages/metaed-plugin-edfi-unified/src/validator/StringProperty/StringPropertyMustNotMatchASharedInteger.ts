import { MetaEdEnvironment, ValidationFailure, ModelBase } from 'metaed-core';
import { getEntityForNamespaces } from 'metaed-core';

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  metaEd.propertyIndex.string.forEach(property => {
    const referencedEntity: ModelBase | null = getEntityForNamespaces(
      property.metaEdName,
      [property.namespace, ...property.namespace.dependencies],
      'sharedInteger',
    );

    if (referencedEntity != null) {
      failures.push({
        validatorName: 'StringPropertyMustNotMatchASharedInteger',
        category: 'error',
        message: `String Property ${property.metaEdName} has the same name as a Shared Integer.`,
        sourceMap: property.sourceMap.metaEdName,
        fileMap: null,
      });
    }
  });

  return failures;
}
