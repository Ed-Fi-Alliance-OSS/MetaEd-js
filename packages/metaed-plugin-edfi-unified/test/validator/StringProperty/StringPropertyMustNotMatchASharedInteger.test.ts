import {
  newMetaEdEnvironment,
  MetaEdTextBuilder,
  DomainEntityBuilder,
  SharedIntegerBuilder,
  NamespaceBuilder,
} from '@edfi/metaed-core';
import { MetaEdEnvironment, ValidationFailure } from '@edfi/metaed-core';
import { validate } from '../../../src/validator/StringProperty/StringPropertyMustNotMatchASharedInteger';

describe('when validating string property does not match shared integer', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const entityName = 'EntityName';
  const maxLength = '10';
  const minLength = '0';

  let failures: ValidationFailure[];
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartSharedInteger('SharedInteger')
      .withDocumentation('doc')
      .withEndSharedInteger()

      .withStartAbstractEntity(entityName, '1')
      .withDocumentation('doc')
      .withStringIdentity('StringIdentity', 'doc', maxLength, minLength)
      .withStringProperty('StringProperty', 'doc', true, false, maxLength, minLength)
      .withEndAbstractEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new SharedIntegerBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build one abstract entity', (): void => {
    expect(coreNamespace.entity.domainEntity.size).toBe(1);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating string identity matches shared integer', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const entityName = 'EntityName';
  const integerProperty = 'IntegerProperty';
  const minLength = '0';
  const maxLength = '10';

  let failures: ValidationFailure[];
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartSharedInteger(integerProperty)
      .withDocumentation('doc')
      .withEndSharedInteger()

      .withStartAbstractEntity(entityName, '1')
      .withDocumentation('doc')
      .withStringIdentity(integerProperty, 'doc', maxLength, minLength)
      .withEndAbstractEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new SharedIntegerBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build one abstract entity', (): void => {
    expect(coreNamespace.entity.domainEntity.size).toBe(1);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('StringPropertyMustNotMatchASharedInteger');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot();
    expect(failures[0].sourceMap).toMatchSnapshot();
  });
});

describe('when validating string property matches shared integer', (): void => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const entityName = 'EntityName';
  const integerProperty = 'IntegerProperty';
  const minLength = '0';
  const maxLength = '10';

  let failures: ValidationFailure[];
  let coreNamespace: any = null;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('EdFi')
      .withStartSharedInteger(integerProperty)
      .withDocumentation('doc')
      .withEndSharedInteger()

      .withStartAbstractEntity(entityName, '1')
      .withDocumentation('doc')
      .withStringProperty(integerProperty, 'doc', true, false, maxLength, minLength)
      .withEndAbstractEntity()
      .withEndNamespace()

      .sendToListener(new NamespaceBuilder(metaEd, []))
      .sendToListener(new DomainEntityBuilder(metaEd, []))
      .sendToListener(new SharedIntegerBuilder(metaEd, []));

    coreNamespace = metaEd.namespace.get('EdFi');
    failures = validate(metaEd);
  });

  it('should build one abstract entity', (): void => {
    expect(coreNamespace.entity.domainEntity.size).toBe(1);
  });

  it('should have no validation failures', (): void => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('StringPropertyMustNotMatchASharedInteger');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot();
    expect(failures[0].sourceMap).toMatchSnapshot();
  });
});
