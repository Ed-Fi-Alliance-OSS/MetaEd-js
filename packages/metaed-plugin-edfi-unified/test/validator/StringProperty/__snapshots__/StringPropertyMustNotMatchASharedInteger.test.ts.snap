// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when validating string identity matches shared integer should have no validation failures 1`] = `"String Property IntegerProperty has the same name as a Shared Integer."`;

exports[`when validating string identity matches shared integer should have no validation failures 2`] = `
Object {
  "column": 11,
  "line": 8,
  "tokenText": "IntegerProperty",
}
`;

exports[`when validating string property matches shared integer should have no validation failures 1`] = `"String Property IntegerProperty has the same name as a Shared Integer."`;

exports[`when validating string property matches shared integer should have no validation failures 2`] = `
Object {
  "column": 11,
  "line": 8,
  "tokenText": "IntegerProperty",
}
`;
